man/awskeyring.5
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "AWSKEYRING" "5" "June 2026" "" ""
.
.SH "NAME"
\fBAwskeyring\fR \- is a small tool to manage AWS account keys in the macOS Keychain
.
.SH "SYNOPSIS"
awskeyring COMMAND [ACCOUNT|ROLE] [OPTIONS]
.
.P
awskeyring help COMMAND
.
.SH "DESCRIPTION"
The Awskeyring utility stores and manages AWS access keys and provides the facility to generate access tokens with combinations of assumed roles and multi\-factor\-authentication codes\. It includes autocompletion features and multiple validation checks for input parsing\. It also includes the ability for the AWS CLI to call it directly to provide authentication\.
.
.P
The commands are as follows:
.
.TP
\-\-version, \-v:
.
.IP
Prints the version
.
.br
.
.IP
\-r, \-\-no\-remote: Do not validate with remote api\.
.
.TP
add ACCOUNT:
.
.IP
Adds an ACCOUNT to the keyring
.
.br
.
.IP
\-k, \-\-key=KEY: AWS account key id\.
.
.br
\-m, \-\-mfa=MFA: AWS virtual mfa arn\.
.
.br
\-r, \-\-no\-remote: Do not validate with remote api\.
.
.br
\-s, \-\-secret=SECRET: AWS account secret\.
.
.TP
add\-role ROLE:
.
.IP
Adds a ROLE to the keyring
.
.br
.
.IP
\-a, \-\-arn=ARN: AWS role arn\.
.
.TP
awskeyring console ACCOUNT:
.
.IP
Open the AWS Console for the ACCOUNT
.
.br
.
.IP
\-b, \-\-browser=BROWSER: Specify an alternative browser\.
.
.br
\-o, \-\-no\-open: Do not open the url\.
.
.br
\-n, \-\-no\-token: Do not use saved token\.
.
.br
\-p, \-\-path=PATH: The service PATH to open\.
.
.TP
env ACCOUNT:
.
.IP
Outputs bourne shell environment exports for an ACCOUNT
.
.br
.
.IP
\-f, \-\-force: Force output to a tty\.
.
.br
\-n, \-\-no\-token: Do not use saved token\.
.
.br
\-t, \-\-test: Generate test credentials\.
.
.br
\-u, \-\-unset: Unset environment variables\.
.
.TP
exec ACCOUNT command\.\.\.:
.
.IP
Execute a COMMAND with the environment set for an ACCOUNT
.
.br
.
.IP
\-b, \-\-no\-bundle: Unset Bundler environment variables\.
.
.br
\-n, \-\-no\-token: Do not use saved token\.
.
.TP
help [COMMAND]:
.
.IP
Describe available commands or one specific command
.
.TP
import:
.
.IP
Import an ACCOUNT to the keyring from ~/\.aws/credentials
.
.br
.
.IP
\-r, \-\-no\-remote: Do not validate with remote api\.
.
.TP
initialise:
.
.IP
Initialises a new KEYCHAIN
.
.br
.
.IP
\-n, \-\-keychain=KEYCHAIN: Name of KEYCHAIN to initialise\.
.
.TP
json ACCOUNT:
.
.IP
Outputs AWS CLI compatible JSON for an ACCOUNT
.
.br
.
.IP
\-f, \-\-force: Force output to a tty\.
.
.br
\-n, \-\-no\-token: Do not use saved token\.
.
.br
\-t, \-\-test: Generate test credentials\.
.
.TP
list:
.
.IP
Prints a list of accounts in the keyring
.
.IP
\-d, \-\-detail, \-\-no\-detail: Show more detail\.
.
.TP
list\-role:
.
.IP
Prints a list of roles in the keyring
.
.br
.
.IP
\-d, \-\-detail, \-\-no\-detail: Show more detail\.
.
.TP
remove ACCOUNT:
.
.IP
Removes an ACCOUNT from the keyring
.
.TP
remove\-role ROLE:
.
.IP
Removes a ROLE from the keyring
.
.TP
remove\-token ACCOUNT:
.
.IP
Removes a token for ACCOUNT from the keyring
.
.TP
rotate ACCOUNT:
.
.IP
Rotate access keys for an ACCOUNT
.
.TP
token ACCOUNT [ROLE] [CODE]:
.
.IP
Create an STS Token from a ROLE or an mfa CODE
.
.br
.
.IP
\-c, \-\-code=CODE: Virtual mfa CODE\.
.
.br
\-d, \-\-duration=DURATION: Session DURATION in seconds\.
.
.TP
update ACCOUNT:
.
.IP
Updates an ACCOUNT in the keyring
.
.br
.
.IP
\-k, \-\-key=KEY: AWS account key id\.
.
.br
\-r, \-\-no\-remote: Do not validate with remote api\.
.
.br
\-s, \-\-secret=SECRET: AWS account secret\.
.
.SH "ENVIRONMENT"
The AWS_DEFAULT_REGION environment variable will be used for AWS API calls where specified or fall back to us\-east\-1 when not\.
.
.SH "EXIT STATUS"
The Awskeyring utility exits 0 on success, and >0 if an error occurs\.
.
.SH "EXAMPLES"
First you need to initialise your keychain to hold your AWS credentials\.
.
.IP "" 4
.
.nf

awskeyring initialise
.
.fi
.
.IP "" 0
.
.P
Then add your keys to it\.
.
.IP "" 4
.
.nf

awskeyring add personal\-aws
.
.fi
.
.IP "" 0
.
.P
Now your keys are stored safely in the macOS keychain\. To print environment variables run\.\.\.
.
.IP "" 4
.
.nf

awskeyring env personal\-aws
.
.fi
.
.IP "" 0
.
.P
To open the AWS Console (web page) with your default browser simply run\.\.\.
.
.IP "" 4
.
.nf

awskeyring console personal\-aws
.
.fi
.
.IP "" 0
.
.P
Autocomplete is enabled in your current shell with the following command\.\.\.
.
.IP "" 4
.
.nf

complete \-C /usr/local/bin/awskeyring awskeyring
.
.fi
.
.IP "" 0
.
.SH "CONFIGURATION"
A Configuration file is stored in the users home directory at \fB~/\.awskeyring\fR as a JSON formatted file\. Most of the fields have a default value except the awskeyring field\.
.
.IP "" 4
.
.nf

{
  "awskeyring": "awskeyring",
  "browser":    ["FireFox", "Google Chrome", "Safari"],
  "console":    ["ec2/v2", "cloudwatch", "iam"],
  "keyage":     90
}
.
.fi
.
.IP "" 0
.
.IP "1." 4
The first field is the Keychain that your keys will be saved in\.
.
.br

.
.IP "2." 4
A list of your browsers to use the console command with\.
.
.br

.
.IP "3." 4
The next is the list of AWS Console pages autocomplete will present\.
.
.br

.
.IP "4." 4
The last field is the warning threshold for key age\.
.
.IP "" 0
.
.SH "HISTORY"
The motivation of this application is to provide a local secure store of AWS credentials using specifically in the macOS Keychain, to have them easily accessed from the Terminal, and to provide useful functions like assuming roles and opening the AWS Console from the cli\. It then expanded to include autocomplete and a desire to have an almost complete test coverage to prevent regressions in its functionality\. For Enterprise environments there are better suited tools to use like HashiCorp Vault \fIhttps://vaultproject\.io/\fR\.
.
.SH "SECURITY"
If you believe you have found a security issue in Awskeyring, please responsibly disclose by contacting me at \fItristan\.morgan@gmail\.com\fR\. Awskeyring is a Ruby script and as such Ruby is whitelisted to access your "awskeyring" keychain\. Use a strong password and keep the unlock time short\.
.
.SH "AUTHOR"
Tristan Morgan \fItristan\.morgan@gmail\.com\fR is the maintainer and author of Awskeyring\.
.
.SH "CONTRIBUTORS"
.
.IP "\(bu" 4
Tristan tristanmorgan \fIhttps://github\.com/tristanmorgan\fR
.
.IP "\(bu" 4
Adam Sir AzySir \fIhttps://github\.com/AzySir\fR
.
.IP "\(bu" 4
Vito Giarrusso thtliife \fIhttps://github\.com/thtliife\fR
.
.IP "" 0
.
.SH "LICENSE"
The gem is available as open source under the terms of the MIT License \fIhttps://opensource\.org/licenses/MIT\fR\.


CHANGELOG.md
# Changelog

## [1.13.0](https://github.com/tristanmorgan/awskeyring/tree/1.13.0) (2025-02-04)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.12.1...1.13.0)

**Merged pull requests:**

- Generate fake creds for testing. [\#94](https://github.com/tristanmorgan/awskeyring/pull/94) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.12.1](https://github.com/tristanmorgan/awskeyring/tree/v1.12.1) (2024-06-04)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.12.0...v1.12.1)

## [v1.12.0](https://github.com/tristanmorgan/awskeyring/tree/v1.12.0) (2023-12-04)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.11.0...v1.12.0)

**Merged pull requests:**

- Decode Account IDs from access keys. [\#93](https://github.com/tristanmorgan/awskeyring/pull/93) ([tristanmorgan](https://github.com/tristanmorgan))
- Migrate away from Travis CI to GitHub Actions [\#92](https://github.com/tristanmorgan/awskeyring/pull/92) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.11.0](https://github.com/tristanmorgan/awskeyring/tree/v1.11.0) (2023-02-21)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.10.1...v1.11.0)

**Implemented enhancements:**

- AWS Access Key Validation tweak [\#91](https://github.com/tristanmorgan/awskeyring/pull/91) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.10.1](https://github.com/tristanmorgan/awskeyring/tree/v1.10.1) (2022-12-12)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.10.0...v1.10.1)

## [v1.10.0](https://github.com/tristanmorgan/awskeyring/tree/v1.10.0) (2022-11-12)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.5...v1.10.0)

**Implemented enhancements:**

- Reorder flags [\#90](https://github.com/tristanmorgan/awskeyring/pull/90) ([tristanmorgan](https://github.com/tristanmorgan))
- Check for tty and allow force. [\#89](https://github.com/tristanmorgan/awskeyring/pull/89) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.5](https://github.com/tristanmorgan/awskeyring/tree/v1.9.5) (2022-10-31)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.4...v1.9.5)

**Implemented enhancements:**

- Exception coverage. [\#88](https://github.com/tristanmorgan/awskeyring/pull/88) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.4](https://github.com/tristanmorgan/awskeyring/tree/v1.9.4) (2022-04-22)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.3...v1.9.4)

**Merged pull requests:**

- ENV.fetch default values. [\#87](https://github.com/tristanmorgan/awskeyring/pull/87) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.3](https://github.com/tristanmorgan/awskeyring/tree/v1.9.3) (2022-01-18)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.2...v1.9.3)

**Implemented enhancements:**

- Stricter aws\_secret validation. [\#86](https://github.com/tristanmorgan/awskeyring/pull/86) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.2](https://github.com/tristanmorgan/awskeyring/tree/v1.9.2) (2021-09-13)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.1...v1.9.2)

**Implemented enhancements:**

- Add CredentialProvider class for library use. [\#85](https://github.com/tristanmorgan/awskeyring/pull/85) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.1](https://github.com/tristanmorgan/awskeyring/tree/v1.9.1) (2021-09-03)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.9.0...v1.9.1)

**Fixed bugs:**

- Install on an M1 Mac \(ARM64\) [\#83](https://github.com/tristanmorgan/awskeyring/issues/83)
- Add account validadtion to exec command. [\#84](https://github.com/tristanmorgan/awskeyring/pull/84) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.9.0](https://github.com/tristanmorgan/awskeyring/tree/v1.9.0) (2021-07-20)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.5...v1.9.0)

**Implemented enhancements:**

- Partial name match for validation. [\#82](https://github.com/tristanmorgan/awskeyring/pull/82) ([tristanmorgan](https://github.com/tristanmorgan))

**Fixed bugs:**

- Complete flags for -v command [\#81](https://github.com/tristanmorgan/awskeyring/pull/81) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.8.5](https://github.com/tristanmorgan/awskeyring/tree/v1.8.5) (2021-06-29)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.4...v1.8.5)

**Merged pull requests:**

- Autocomplete for exec and long flags [\#80](https://github.com/tristanmorgan/awskeyring/pull/80) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.8.4](https://github.com/tristanmorgan/awskeyring/tree/v1.8.4) (2021-05-24)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.3...v1.8.4)

**Merged pull requests:**

- \(Auto\)Complete rewrite [\#79](https://github.com/tristanmorgan/awskeyring/pull/79) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.8.3](https://github.com/tristanmorgan/awskeyring/tree/v1.8.3) (2021-04-26)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.2...v1.8.3)

**Implemented enhancements:**

- Package name and default action [\#78](https://github.com/tristanmorgan/awskeyring/pull/78) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.8.2](https://github.com/tristanmorgan/awskeyring/tree/v1.8.2) (2021-03-19)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.1...v1.8.2)

**Implemented enhancements:**

- Make autocomplete accept dynamic exec\_name [\#77](https://github.com/tristanmorgan/awskeyring/pull/77) ([tristanmorgan](https://github.com/tristanmorgan))

## [v1.8.1](https://github.com/tristanmorgan/awskeyring/tree/v1.8.1) (2021-01-20)

[Full Changelog](https://github.com/tristanmorgan/awskeyring/compare/v1.8.0...v1.8.1)

**Fixed bugs:**


lib/awskeyring.rb
# frozen_string_literal: true

require 'i18n'
require 'json'
require 'keychain'
require 'awskeyring/validate'
require 'awskeyring/awsapi'

# Awskeyring Module,
# gives you an interface to access keychains and items.
module Awskeyring # rubocop:disable Metrics/ModuleLength
  I18n.load_path = Dir.glob(File.join(File.realpath(__dir__), '..', 'i18n', '*.{yml,yaml}'))
  I18n.backend.load_translations

  # Default rpeferences fole path
  PREFS_FILE = (File.expand_path '~/.awskeyring').freeze
  # Prefix for Roles
  ROLE_PREFIX = 'role '
  # Prefix for Accounts
  ACCOUNT_PREFIX = 'account '
  # Prefix for Session Keys
  SESSION_KEY_PREFIX = 'session-key '
  # Prefix for Session Tokens
  SESSION_TOKEN_PREFIX = 'session-token '
  # Default keychain Lock period
  FIVE_MINUTES = 300
  # Default warning of key age in days.
  DEFAULT_KEY_AGE = 90
  # Default Console Paths
  DEFAULT_CONSOLE_LIST = %w[cloudformation ec2/v2 iam rds route53 s3 sns sqs vpc].freeze
  # Default Browsers
  DEFAULT_BROWSER_LIST = %w[Brave FireFox Opera Safari Vivaldi].freeze

  # Retrieve the preferences
  #
  # @return [Hash] prefs of the gem
  def self.prefs
    if File.exist? PREFS_FILE
      JSON.parse(File.read(PREFS_FILE))
    else
      {}
    end
  end

  # Create a new Keychain
  #
  # @param [String] awskeyring The keychain name to create
  def self.init_keychain(awskeyring:)
    keychain = Keychain.create(awskeyring)
    keychain.lock_interval = FIVE_MINUTES
    keychain.lock_on_sleep = true

    prefs = {
      awskeyring: awskeyring,
      keyage: DEFAULT_KEY_AGE,
      browser: DEFAULT_BROWSER_LIST,
      console: DEFAULT_CONSOLE_LIST
    }
    File.new(Awskeyring::PREFS_FILE, 'w').write JSON.dump(prefs)
  end

  # Load the keychain for access
  #
  # @return [Keychain] keychain ready for use.
  private_class_method def self.load_keychain
    unless File.exist?(Awskeyring::PREFS_FILE) && !prefs.empty?
      warn I18n.t('message.missing', bin: File.basename($PROGRAM_NAME))
      exit 1
    end

    keychain = Keychain.open(prefs['awskeyring'])
    warn I18n.t('message.timeout') if keychain && keychain.lock_interval > FIVE_MINUTES

    keychain
  end

  # Return a list of all acount items
  private_class_method def self.list_items
    all_items.all.select { |elem| elem.attributes[:label].start_with?(ACCOUNT_PREFIX) }
  end

  # Return a list of all role items
  private_class_method def self.list_roles
    all_items.all.select { |elem| elem.attributes[:label].start_with?(ROLE_PREFIX) }
  end

  # Return a list of all acount items
  private_class_method def self.list_tokens
    all_items.all.select { |elem| elem.attributes[:label].start_with?(SESSION_KEY_PREFIX) }
  end

  # Return all keychain items
  private_class_method def self.all_items
    load_keychain.generic_passwords
  end

  # return an item by accout
  private_class_method def self.item_by_account(account)
    all_items.where(account: account).first
  end

  # return item that matches a prefix if only one.
  def self.solo_select(list, prefix)
    return prefix if list.include?(prefix)

    list.select! { |elem| elem.start_with?(prefix) }

    return list.first if list.length == 1

    nil
  end

  # Add an account item
  #
  # @param [String] account The account name to create
  # @param [String] key The aws_access_key_id
  # @param [String] secret The aws_secret_key
  # @param [String] mfa The arn of the MFA device
  def self.add_account(account:, key:, secret:, mfa:)
    all_items.create(
      label: ACCOUNT_PREFIX + account,
      account: key,
      password: secret,
      comment: mfa
    )
  end

  # update and account item
  #
  # @param [String] account The account to update
  # @param [String] key The aws_access_key_id
  # @param [String] secret The aws_secret_key
  def self.update_account(account:, key:, secret:)
    item = get_item(account: account)
    item.attributes[:account] = key
    item.password = secret
    item.save!
  end

  # Add a Role item
  #
  # @param [String] role The role name to add
  # @param [String] arn The arn of the role
  def self.add_role(role:, arn:)
    all_items.create(
      label: ROLE_PREFIX + role,
      account: arn,
      password: '',
      comment: ''
    )
  end

  # add a session token pair of items
  #
  # @param [Hash] params including
  #    account The name of the accont
  #    key The aws_access_key_id
  #    secret The aws_secret_access_key
  #    token The aws_sesson_token
  #    expiry time of expiry
  #    role The role used
  def self.add_token(params = {})
    all_items.create(label: SESSION_KEY_PREFIX + params[:account],
                     account: params[:key],
                     password: params[:secret],
                     comment: params[:role].nil? ? '' : ROLE_PREFIX + params[:role])
    all_items.create(label: SESSION_TOKEN_PREFIX + params[:account],
                     account: params[:expiry],
                     password: params[:token],
                     comment: params[:role] || '')
  end

  # Return an account item by name
  private_class_method def self.get_item(account:)
    all_items.where(label: ACCOUNT_PREFIX + account).first
  end

  # Return a role item by name
  private_class_method def self.get_role(role_name:)
    all_items.where(label: ROLE_PREFIX + role_name).first
  end

  # Return a session token pair of items by name
  private_class_method def self.get_token_pair(account:)
    session_key = all_items.where(label: SESSION_KEY_PREFIX + account).first
    session_token = all_items.where(label: SESSION_TOKEN_PREFIX + account).first if session_key
    [session_key, session_token]
  end

  # Return a list account item names
  def self.list_account_names
    items = list_items.map { |elem| elem.attributes[:label][(ACCOUNT_PREFIX.length)..] }

    tokens = list_tokens.map { |elem| elem.attributes[:label][(SESSION_KEY_PREFIX.length)..] }

    (items + tokens).uniq.sort
  end

  # Return a list account item names plus account ids
  def self.list_account_names_plus # rubocop:disable Metrics/AbcSize
    list_items.concat(list_tokens).map do |elem|
      account_id = Awskeyring::Awsapi.get_account_id(key: elem.attributes[:account])
      account_name = if elem.attributes[:label].start_with?(ACCOUNT_PREFIX)
                       elem.attributes[:label][(ACCOUNT_PREFIX.length)..]
                     else
                       elem.attributes[:label][(SESSION_KEY_PREFIX.length)..]
                     end
      "#{account_name}\t#{account_id}"
    end.uniq.sort
  end

  # Return a list role item names
  def self.list_role_names
    list_roles.map { |elem| elem.attributes[:label][(ROLE_PREFIX.length)..] }.sort
  end

  # Return a list token item names
  def self.list_token_names
    list_tokens.map { |elem| elem.attributes[:label][(SESSION_KEY_PREFIX.length)..] }.sort
  end

  # Return a list role item names and arns
  def self.list_role_names_plus
    list_roles.map { |elem| "#{elem.attributes[:label][(ROLE_PREFIX.length)..]}\t#{elem.attributes[:account]}" }
  end

  # Return a list of console paths
  def self.list_console_path
    prefs.key?('console') ? prefs['console'] : DEFAULT_CONSOLE_LIST
  end

  # Return a list of browserss
  def self.list_browsers
    prefs.key?('browser') ? prefs['browser'] : DEFAULT_BROWSER_LIST
  end

  # Return Key age warning number
  def self.key_age
    prefs.key?('keyage') ? prefs['keyage'] : DEFAULT_KEY_AGE
  end

  # Return a session token if available or a static key
  private_class_method def self.get_valid_item_pair(account:, no_token: false)
    session_key, session_token = get_token_pair(account: account)
    session_key, session_token = delete_expired(key: session_key, token: session_token) if session_key

    if session_key && session_token && !no_token
      puts I18n.t('message.temporary')
      return session_key, session_token
    end

    item = get_item(account: account)
    if item.nil?
      warn I18n.t('message.notfound', account: account)
      exit 2
    end
    [item, nil]
  end

  # Return valid creds for account
  #
  # @param [String] account The account to retrieve
  # @param [Boolean] no_token Flag to skip tokens
  def self.get_valid_creds(account:, no_token: false)
    cred, temp_cred = get_valid_item_pair(account: account, no_token: no_token)
    token = temp_cred.password unless temp_cred.nil?
    expiry = temp_cred.attributes[:account].to_i unless temp_cred.nil?
    {
      account: account,
      expiry: expiry,
      key: cred.attributes[:account],
      mfa: no_token ? cred.attributes[:comment] : nil,
      secret: cred.password,
      token: token,
      updated: cred.attributes[:updated_at]
    }
  end

  # get the ARN for a role
  #
  # @param [String] role_name The role name to retrieve
  def self.get_role_arn(role_name:)
    role_item = get_role(role_name: role_name)
    role_item.attributes[:account] if role_item
  end

  # Delete session token items if expired
  private_class_method def self.delete_expired(key:, token:)
    expires_at = Time.at(token.attributes[:account].to_i)
    if expires_at < Time.new
      delete_pair(key: key, token: token, message: I18n.t('message.delexpired'))
      key = nil
      token = nil
    end
    [key, token]
  end

  # Delete session token items
  private_class_method def self.delete_pair(key:, token:, message:)
    return unless key

    puts message if message
    token.delete if token
    key.delete
  end

  # Delete a session token
  #
  # @param [String] account The account to delete a token for
  # @param [String] message The message to display
  def self.delete_token(account:, message:)
    session_key, session_token = get_token_pair(account: account)
    delete_pair(key: session_key, token: session_token, message: message)
  end

  # Delete an Account
  #
  # @param [String] account The account to delete
  # @param [String] message The message to display
  def self.delete_account(account:, message:)
    delete_token(account: account, message: I18n.t('message.delexpired'))
    cred = get_item(account: account)
    return unless cred

    puts message if message
    cred.delete
  end

  # Delete a role
  #
  # @param [String] role_name The role to delete
  # @param [String] message The message to display
  def self.delete_role(role_name:, message:)
    role = get_role(role_name: role_name)
    return unless role

    puts message if message
    role.delete
  end

  # Validate account exists
  #
  # @param [String] account_name the associated account name.
  def self.account_exists(account_name)
    Awskeyring::Validate.account_name(account_name)
    raise 'Account does not exist' unless (account_name = solo_select(list_account_names, account_name))

    account_name
  end

  # Validate account does not exists
  #
  # @param [String] account_name the associated account name.
  def self.account_not_exists(account_name)
    Awskeyring::Validate.account_name(account_name)
    raise 'Account already exists' if list_account_names.include?(account_name)

    account_name
  end

  # Validate access key does not exists
  #
  # @param [String] access_key the associated access key.
  def self.access_key_not_exists(access_key)
    Awskeyring::Validate.access_key(access_key)
    raise 'Access KEY already exists' if item_by_account(access_key)

    access_key
  end

  # Validate role exists
  #
  # @param [String] role_name the associated role name.
  def self.role_exists(role_name)
    Awskeyring::Validate.role_name(role_name)
    raise 'Role does not exist' unless (role_name = solo_select(list_role_names, role_name))

    role_name
  end

  # Validate role does not exists
  #
  # @param [String] role_name the associated role name.
  def self.role_not_exists(role_name)
    Awskeyring::Validate.role_name(role_name)
    raise 'Role already exists' if list_role_names.include?(role_name)

    role_name
  end

  # Validate token exists
  #
  # @param [String] token_name the associated account name.
  def self.token_exists(token_name)
    Awskeyring::Validate.account_name(token_name)
    raise 'Token does not exist' unless (token_name = solo_select(list_token_names, token_name))

    token_name
  end

  # Validate role arn not exists
  #
  # @param [String] role_arn the associated role arn.
  def self.role_arn_not_exists(role_arn)
    Awskeyring::Validate.role_arn(role_arn)
    raise 'Role ARN already exists' if item_by_account(role_arn)

    role_arn
  end
end
